File: fwdpy11/src/fwdpy11_types/DiploidPopulation.cc
//
// Copyright (C) 2017 Kevin Thornton <krthornt@uci.edu>
//
// This file is part of fwdpy11.
//
// fwdpy11 is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// fwdpy11 is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with fwdpy11.  If not, see <http://www.gnu.org/licenses/>.
//

// TODO: make a versionchanged entry for all things affected by "length"

#include <limits>
#include <sstream>
#include <fstream>
#include <type_traits>
#include <pybind11/pybind11.h>
#include <pybind11/stl.h>
#include <fwdpy11/types/DiploidPopulation.hpp>
#include <fwdpy11/types/create_pops.hpp>
#include <fwdpy11/serialization.hpp>
#include <fwdpy11/serialization/Mutation.hpp>
#include <fwdpy11/serialization/Diploid.hpp>
#include "get_individuals.hpp"

namespace py = pybind11;

namespace
{
    static const auto DIPLOIDS_DOCSTRING = R"delim(
   A :class:`fwdpy11.DiploidVector`.
   )delim";

} // namespace

PYBIND11_MAKE_OPAQUE(std::vector<fwdpy11::DiploidGenotype>);
PYBIND11_MAKE_OPAQUE(fwdpy11::DiploidPopulation::gcont_t);
PYBIND11_MAKE_OPAQUE(fwdpy11::DiploidPopulation::mcont_t);

fwdpy11::DiploidPopulation
create_DiploidPopulation_from_tree_sequence(py::object ts);

void
init_DiploidPopulation(py::module& m)
{
    py::class_<fwdpy11::DiploidPopulation, fwdpy11::Population>(
        m, "DiploidPopulation", "Representation of a diploid population")
        .def(py::init<fwdpp::uint_t, double>(),
             "Construct with an unsigned integer "
             "representing the initial "
             "population size.",
             py::arg("N"),
             py::arg("length") = std::numeric_limits<double>::max())
        .def(py::init<const fwdpy11::DiploidPopulation::dipvector_t&,
                      const fwdpy11::DiploidPopulation::gcont_t&,
                      const fwdpy11::DiploidPopulation::mcont_t&>(),
             R"delim(
             Construct with tuple of (diploids, haploid_genomes, mutations).
             
             .. versionadded:: 0.1.4
             )delim",
             py::arg("diploids"), py::arg("haploid_genomes"),
             py::arg("mutations"))
        .def(py::init<const fwdpy11::DiploidPopulation&>(),
             R"delim(
                Copy constructor

                .. versionadded:: 0.1.4
                )delim")
        .def("clear", &fwdpy11::DiploidPopulation::clear,
             "Clears all population data.")
        .def("__eq__",
             [](const fwdpy11::DiploidPopulation& lhs,
                const fwdpy11::DiploidPopulation& rhs) { return lhs == rhs; })
        .def_readonly("diploids", &fwdpy11::DiploidPopulation::diploids,
                      DIPLOIDS_DOCSTRING)
        .def_static(
            "create",
            [](fwdpy11::DiploidPopulation::dipvector_t& diploids,
               fwdpy11::DiploidPopulation::gcont_t& haploid_genomes,
               fwdpy11::DiploidPopulation::mcont_t& mutations,
               py::args args) -> fwdpy11::DiploidPopulation {
                if (args.size() == 0)
                    {
                        return fwdpy11::create_wrapper<
                            fwdpy11::DiploidPopulation>()(
                            std::move(diploids), std::move(haploid_genomes),
                            std::move(mutations));
                    }
                auto& fixations
                    = args[0].cast<fwdpy11::DiploidPopulation::mcont_t&>();
                auto ftimes = args[1].cast<std::vector<fwdpp::uint_t>>();
                auto g = args[2].cast<fwdpp::uint_t>();
                return fwdpy11::create_wrapper<fwdpy11::DiploidPopulation>()(
                    std::move(diploids), std::move(haploid_genomes),
                    std::move(mutations), std::move(fixations),
                    std::move(ftimes), g);
            },
            py::arg("diploids"), py::arg("haploid_genomes"),
            py::arg("mutations"),
            R"delim(
        Create a new object from input data.
        Unlike the constructor method, this method results
        in no temporary copies of input data.

        :param diplods: A :class:`fwdpy11.VecDiploid`
        :param haploid_genomes: A :class:`fwdpy11.VecGamete`
        :param mutations: A :class:`fwdpy11.VecMutation`
        :param args: Fixations, fixation times, and generation

        :rtype: :class:`fwdpy11.DiploidPopulation`

        See :ref:`popobjects` for example use.

        When passing in extra args, they must be the following:

        fixations: A :class:`fwdpy11.VecMutation`
        fixation times: A :class:`fwdpy11.VecUint32`
        generation: A non-negative integer

        It is required that len(fixations) == len(fixation times).

        The result of passing in these extra args will be an object
        with its fixation data populated and its generation set
        to the input value.

        .. versionadded:: 0.1.4
        )delim")
        .def(py::pickle(
            [](const fwdpy11::DiploidPopulation& pop) -> py::object {
                std::ostringstream o;
                fwdpy11::serialization::serialize_details(o, &pop);
                auto pb = py::bytes(o.str());
                return py::object(std::move(pb));
            },
            [](py::object pickled) -> fwdpy11::DiploidPopulation {
                auto s = pickled.cast<py::bytes>().cast<std::string>();
                fwdpy11::DiploidPopulation pop(
                    1, std::numeric_limits<double>::max());
                std::istringstream in(std::move(s));
                fwdpy11::serialization::deserialize_details()(in, pop);
                return pop;
            }))
        .def(
            "sample",
            [](const fwdpy11::DiploidPopulation& pop,
               const std::vector<std::size_t>& individuals,
               const bool haplotype, const bool remove_fixed) {
                return pop.sample_individuals(individuals, haplotype,
                                              remove_fixed);
            },
            R"delim(
             Return a sample from a population.

             :param individuals: Indexes of individuals in the sample
             :type individuals: list
             :param haplotype: (True) Determines the encoding of the return type
             :type haplotype: bool
             :param remove_fixed: (True) Remove fixed variants from the sample
             :type remove_fixed: bool

             :returns: A haplotype matrix if haplotype is True. Otherwise, a genotype matrix.
             :rtype: :class:`fwdpy11.sampling.DataMatrix`

             .. versionadded:: 0.1.4

             .. versionchanged:: 0.2.0

                Change to an overloaded function returning a 
                :class:`fwdpy11.sampling.DataMatrix` instead of 
                the "classic libsequence" layout.
             )delim",
            py::arg("individuals"), py::arg("haplotype") = true,
            py::arg("remove_fixed") = true)
        .def(
            "sample",
            [](const fwdpy11::DiploidPopulation& pop,
               const fwdpy11::GSLrng_t& rng, const std::uint32_t nsam,
               const bool haplotype, const bool remove_fixed) {
                return pop.sample_random_individuals(rng, nsam, haplotype,
                                                     remove_fixed);
            },
            py::arg("rng"), py::arg("nsam"), py::arg("haplotype") = true,
            py::arg("remove_fixed") = true,
            R"delim(
             Return a sample from a population.

             :param rng: Random number generator
             :type rng: :class:`fwdpy11.GSLrng`
             :param nsam: The sample size
             :type nsam: int
             :param haplotype: (True) Determines the encoding of the return type
             :type haplotype: bool
             :param remove_fixed: (True) Remove fixed variants from the sample
             :type remove_fixed: bool

             :returns: A haplotype matrix if haplotype is True. Otherwise, a genotype matrix.
             :rtype: :class:`fwdpy11.sampling.DataMatrix`

             .. versionadded:: 0.1.4

             .. versionchanged:: 0.2.0

                Change to an overloaded function returning a 
                :class:`fwdpy11.sampling.DataMatrix` instead of 
                the "classic libsequence" layout.
             )delim")
        .def("add_mutations", &fwdpy11::DiploidPopulation::add_mutations)
        .def(
            "dump_to_file",
            [](const fwdpy11::DiploidPopulation& pop,
               const std::string filename) {
                std::ofstream out(filename.c_str(), std::ios_base::binary);
                if (!out)
                    {
                        throw std::runtime_error(
                            "could not open file for writing");
                    }
                fwdpy11::serialization::serialize_details(out, &pop);
                out.close();
            },
            "Write a population to a file in binary format.")
        .def_static(
            "load_from_file",
            [](const std::string filename) {
                std::ifstream in(filename.c_str(), std::ios_base::binary);
                if (!in)
                    {
                        throw std::runtime_error(
                            "could not open file for reading");
                    }
                fwdpy11::DiploidPopulation pop(
                    1, std::numeric_limits<double>::max());
                fwdpy11::serialization::deserialize_details()(in, pop);
                return pop;
            },
            "Load a population from a binary file.")
        .def(
            "pickle_to_file",
            [](const fwdpy11::DiploidPopulation& self, py::object f) {
                auto dump = py::module::import("pickle").attr("dump");
                dump(py::make_tuple(
                         self.diploids.size(), self.haploid_genomes.size(),
                         self.mutations.size(), self.fixations.size(),
                         self.generation, self.tables.genome_length()),
                     f);
                for (auto& d : self.diploids)
                    {
                        dump(d, f);
                    }
                for (auto& g : self.haploid_genomes)
                    {
                        dump(g, f);
                    }
                for (auto& m : self.mutations)
                    {
                        dump(m, f);
                    }
                for (auto& m : self.fixations)
                    {
                        dump(m, f);
                    }
                dump(self.fixation_times, f);
                dump(self.mcounts, f);
                dump(self.mcounts_from_preserved_nodes, f);
                dump(py::make_tuple(self.diploid_metadata.size(),
                                    self.ancient_sample_metadata.size()),
                     f);
                for (auto& md : self.diploid_metadata)
                    {
                        dump(md, f);
                    }
                for (auto& md : self.ancient_sample_metadata)
                    {
                        dump(md, f);
                    }
                dump(py::make_tuple(self.tables.node_table.size(),
                                    self.tables.edge_table.size(),
                                    self.tables.mutation_table.size(),
                                    self.tables.site_table.size()),
                     f);
                for (auto& n : self.tables.node_table)
                    {
                        dump(n, f);
                    }
                for (auto& e : self.tables.edge_table)
                    {
                        dump(e, f);
                    }
                for (auto& m : self.tables.mutation_table)
                    {
                        dump(m, f);
                    }
                for (auto& s : self.tables.site_table)
                    {
                        dump(s, f);
                    }
                dump(self.tables.preserved_nodes, f);
            },
            R"delim(
             Pickle the population to an open file.

             This function may be preferred over 
             the direct pickling method because it uses less
             memory.  It is, however, slower.

             To read the population back in, you must call
             :func:`fwdpy11.DiploidPopulation.load_from_pickle_file`.

             :param f: A handle to an open file

             .. versionadded:: 0.3.0
             )delim")
        .def_static(
            "load_from_pickle_file",
            [](py::object f) {
                auto load = py::module::import("pickle").attr("load");
                py::tuple popdata = load(f);
                fwdpy11::DiploidPopulation rv(popdata[0].cast<fwdpp::uint_t>(),
                                              popdata[5].cast<double>());
                rv.generation
                    = popdata[4]
                          .cast<decltype(
                              fwdpy11::DiploidPopulation::generation)>();
                auto ndips = popdata[0].cast<std::size_t>();
                auto ngams = popdata[1].cast<std::size_t>();
                auto nmuts = popdata[2].cast<std::size_t>();
                auto nfixations = popdata[3].cast<std::size_t>();
                rv.diploids.clear();
                rv.haploid_genomes.clear();
                rv.mutations.clear();
                rv.fixations.clear();
                rv.diploids.reserve(ndips);
                for (std::size_t i = 0; i < ndips; ++i)
                    {
                        rv.diploids.push_back(
                            load(f).cast<fwdpy11::DiploidGenotype>());
                    }
                rv.haploid_genomes.reserve(ngams);
                for (std::size_t i = 0; i < ngams; ++i)
                    {
                        rv.haploid_genomes.push_back(
                            load(f).cast<fwdpp::haploid_genome>());
                    }
                rv.mutations.reserve(nmuts);
                for (std::size_t i = 0; i < nmuts; ++i)
                    {
                        rv.mutations.push_back(
                            load(f).cast<fwdpy11::Mutation>());
                    }
                rv.fixations.reserve(nfixations);
                for (std::size_t i = 0; i < nfixations; ++i)
                    {
                        rv.fixations.push_back(
                            load(f).cast<fwdpy11::Mutation>());
                    }
                rv.fixation_times
                    = load(f).cast<decltype(rv.fixation_times)>();
                rv.mcounts = load(f).cast<decltype(rv.mcounts)>();
                rv.mcounts_from_preserved_nodes
                    = load(f)
                          .cast<decltype(rv.mcounts_from_preserved_nodes)>();
                py::tuple metadata_data = load(f);
                rv.diploid_metadata.clear();
                rv.ancient_sample_metadata.clear();
                auto lmd = metadata_data[0].cast<std::size_t>();
                auto lamd = metadata_data[1].cast<std::size_t>();
                rv.diploid_metadata.reserve(lmd);
                for (std::size_t i = 0; i < lmd; ++i)
                    {
                        rv.diploid_metadata.push_back(
                            load(f).cast<fwdpy11::DiploidMetadata>());
                    }
                rv.ancient_sample_metadata.reserve(lamd);
                for (std::size_t i = 0; i < lamd; ++i)
                    {
                        rv.ancient_sample_metadata.push_back(
                            load(f).cast<fwdpy11::DiploidMetadata>());
                    }
                py::tuple table_data = load(f);
                auto table_len = table_data[0].cast<std::size_t>();
                rv.tables.clear();
                rv.tables.node_table.reserve(table_len);
                for (std::size_t i = 0; i < table_len; ++i)
                    {
                        rv.tables.node_table.push_back(
                            load(f).cast<fwdpp::ts::node>());
                    }
                table_len = table_data[1].cast<std::size_t>();
                rv.tables.edge_table.reserve(table_len);
                for (std::size_t i = 0; i < table_len; ++i)
                    {
                        rv.tables.edge_table.push_back(
                            load(f).cast<fwdpp::ts::edge>());
                    }
                table_len = table_data[2].cast<std::size_t>();
                rv.tables.mutation_table.reserve(table_len);
                for (std::size_t i = 0; i < table_len; ++i)
                    {
                        rv.tables.mutation_table.push_back(
                            load(f).cast<fwdpp::ts::mutation_record>());
                    }
                table_len = table_data[3].cast<std::size_t>();
                rv.tables.site_table.reserve(table_len);
                for (std::size_t i = 0; i < table_len; ++i)
                    {
                        rv.tables.site_table.push_back(
                            load(f).cast<fwdpp::ts::site>());
                    }

                rv.tables.preserved_nodes
                    = load(f).cast<decltype(rv.tables.preserved_nodes)>();
                rv.tables.build_indexes();
                return rv;
            },
            R"delim(
            Read in a pickled population from a file.
            The file muse have been generated by
            a call to :func:`fwdpy11.DiploidPopulation.pickle_to_file`.

            :param f: A handle to a file opened in 'rb' mode.

            .. versionadded: 0.3.0
            )delim")
        .def_static(
            "create_from_tskit",
            [](py::object ts) {
                return create_DiploidPopulation_from_tree_sequence(ts);
            },
            R"delim(
        Create a new object from an tskit.TreeSequence

        :param ts: A tree sequence from tskit
        :type ts: tskit.TreeSequence

        :rtype: :class:`fwdpy11.DiploidPopulation`
        :returns: A population object with an initialized
        :class:`fwdpy11.TableCollection`

        .. versionadded:: 0.2.0

        .. note::

            In general, initializing a population using
            the output from a coalescent simulation is
            a tricky business.  There are issues of
            parameter scaling and the appropriateness
            of the coalescent model itself. A key issue
            is that your input tree sequence must have
            node times in the correct time units! (Generations,
            for example.) See :ref:`ts` for more discussion
        )delim");
}

File: fwdpy11/_types/diploid_population.py
from typing import IO, Dict, Iterable, Iterator, List, Optional, Tuple, Union

import demes
import fwdpy11._types
import fwdpy11.tskit_tools._dump_tables_to_tskit
import numpy as np
import tskit  # type: ignore

from .._fwdpy11 import DiploidGenotype, DiploidMetadata, ll_DiploidPopulation
from .model_params import ModelParams
from .new_mutation_data import NewMutationData
from .population_mixin import PopulationMixin
from .table_collection import TableCollection


class DiploidPopulation(ll_DiploidPopulation, PopulationMixin):
    """
    A population of diploid individuals.

    To initialize a population of size N and genome length L:

    fwdpy11.DiploidPopulation(N, L)

    To initialize a population with two demes of initial
    sizes N1 and N2 and genome length L:

    fwdpy11.DiploidPopulation([N1, N2], L)

    .. versionchanged:: 0.16.0

        Added __copy__ and __deepcopy__.
        In general, the correct way to copy
        instances if this class is via :func:`copy.deepcopy`.
        Calling :func:`copy.copy` will copy some of the underlying
        C++ objects, but the two objects will share the same
        table collection.  This situation of sharing is almost certainly
        not what one wants.
    """

    def __init__(
        self,
        N: Union[int, List[int]],
        length: float,
        *,
        ll_pop: Optional[ll_DiploidPopulation] = None,
    ):
        if ll_pop is None:
            super(DiploidPopulation, self).__init__(N, length)
        else:
            super(DiploidPopulation, self).__init__(ll_pop)

        self._pytables = TableCollection(self._tables)

    def __copy__(self):
        ll_pop = super(DiploidPopulation, self).__copy__()
        return self.__class__(0, 0.0, ll_pop=ll_pop)

    def __deepcopy__(self, memo):
        ll_pop = super(DiploidPopulation, self).__deepcopy__(memo)
        return self.__class__(0, 0.0, ll_pop=ll_pop)

    @classmethod
    def create_from_tskit(cls, ts: tskit.TreeSequence,
                          *, import_mutations=False):
        """
        Create a new object from an tskit.TreeSequence

        :param ts: A tree sequence from tskit
        :type ts: tskit.TreeSequence
        :param import_mutations: if `True`, import mutations and sites
                                 from the tree sequence
        :type import_mutations: bool

        :return: A population object with an initialized
                 :class:`fwdpy11.TableCollection`
        :rtype: :class:`fwdpy11.DiploidPopulation`

        .. note::

            In general, initializing a population using
            the output from a coalescent simulation is
            a tricky business.  There are issues of
            parameter scaling and the appropriateness
            of the coalescent model itself. A key issue
            is that your input tree sequence must have
            node times in the correct time units! (Generations,
            for example.)

        .. warning::

            Support for importing mutations is currently limited to
            two cases:

            1. The tree seq has mutations entirely from tskit's API
               (see :ref:`import_mutations_from_tskit_vignette`).
            2. The tree seq has mutations entirely from a previous run
               of fwdpy11.


        .. versionadded:: 0.2.0

        .. versionchanged:: 0.19.8

           Correctly handle two cases:

           1. The tree seq has mutations entirely from tskit's API
              (see :ref:`import_mutations_from_tskit_vignette`).
           2. The tree seq has mutations entirely from a previous run
              of fwdpy11.

        """
        generation = 0
        if len(ts.metadata) > 0:
            if "generation" in ts.metadata:
                generation = ts.metadata["generation"]
        ll = ll_DiploidPopulation._create_from_tskit(ts, generation)
        if import_mutations is True:
            import fwdpy11.tskit_tools
            for mutation in ts.mutations():
                # if mutation.metadata["origin"] < 0.0:
                #     msg = "all origin fields in mutation metadata"
                #     msg += "must be non-negative"
                #     raise ValueError(msg)
                if not mutation.time.is_integer():
                    raise ValueError("mutation times cannot contain decimals")
            mutation_nodes = [i.node for i in ts.mutations()]
            decoded_md = fwdpy11.tskit_tools.decode_mutation_metadata(ts)
            try:
                # The case of the tree seq has all mutations added
                # to it via tskit API
                for i, j in zip(ts.mutations(), decoded_md):
                    if j is not None:
                        if i.time != j.g:
                            msg = f"metadata origin time {j.g} != "
                            msg += f"mutation time {i.time} in tables"
                            raise ValueError(msg)
            except ValueError:
                # The case of the tree seq has all of its mutations
                # from a pervious fwdpy11 run
                keys = set([i.key for i in decoded_md])
                for i in ts.mutations():
                    md = i.metadata
                    key = (ts.site(i.site).position, md['s'], md['origin'])
                    if key not in keys:
                        raise ValueError(f"{key}")

            mvec = fwdpy11._fwdpy11.MutationVector()
            for i in decoded_md:
                if i is not None:
                    mvec.append(i)
            ll._set_mutations(mvec, mutation_nodes,
                              [int(i.time) - ll._generation for i in ts.mutations()])
        return cls(0, 0.0, ll_pop=ll)

    @ classmethod
    def load_from_file(cls, filename: str):
        """
        Load population from the output of
        :func:`fwdpy11.DiploidPopulation.dump_to_file`.

        :param filename: A file name
        :type filename: str
        """
        ll = ll_DiploidPopulation._load_from_file(filename)
        return cls(0, 0.0, ll_pop=ll)

    @ classmethod
    def load_from_pickle_file(cls, filename: IO):
        """
        Read in a pickled population from a file.
        The file muse have been generated by
        a call to :func:`fwdpy11.DiploidPopulation.pickle_to_file`.

        :param f: A handle to a file opened in 'rb' mode.

        .. versionadded: 0.3.0

        """
        ll = ll_DiploidPopulation._load_from_pickle_file(filename)
        return cls(0, 0.0, ll_pop=ll)

    @ property
    def alive_nodes(self) -> np.ndarray:
        """
        List of alive nodes corresponding to individuals.

        .. versionadded:: 0.5.3
        """
        md = np.array(self.diploid_metadata, copy=False)
        return md["nodes"].flatten()

    @ property
    def ancient_sample_nodes(self) -> np.ndarray:
        """
        Return an array of nodes associated with preserved/ancient samples.

        Alias for :attr:`fwdpy11.DiploidPopulation.preserved_nodes`.

        .. versionadded:: 0.13.0
        """
        return self.preserved_nodes

    @ property
    def ancient_sample_metadata(self) -> Iterable[DiploidMetadata]:
        """Supports buffer protocol"""
        return self._ancient_sample_metadata

    @ property
    def diploids(self) -> Iterable[DiploidGenotype]:
        """Supports buffer protocol"""
        return self._diploids

    @ property
    def diploid_metadata(self) -> Iterable[DiploidMetadata]:
        """Supports buffer protocol"""
        return self._diploid_metadata

    @ property
    def preserved_nodes(self) -> np.ndarray:
        """
        Return an array of nodes associated with preserved/ancient samples.

        :rtype: numpy.ndarray

        .. versionadded:: 0.13.0

        """
        return np.array(self.ancient_sample_metadata, copy=False)["nodes"].flatten()

    @ property
    def tables(self) -> TableCollection:
        """Access the :class:`fwdpy11.TableCollection`"""
        return self._pytables

    def _get_times(self):
        amd = np.array(self.ancient_sample_metadata, copy=False)
        nodes = np.array(self.tables.nodes, copy=False)
        times = nodes["time"][amd["nodes"][:, 0]]
        utimes = np.unique(times)
        return times, utimes

    def deme_sizes(self, as_dict=False):
        """
        Return the number of individuals in each deme.

        :param as_dict: If True, return results as a dict.
                        If False, numpy arrays are returned

        The default behavior of this function is equivalent to:

        .. code-block:: python

            md = numpy.array(self.diploid_metadata, copy=False)
            return numpy.unique(md['deme'], return_counts=True)

        .. versionadded:: 0.6.0
        """
        md = np.array(self.diploid_metadata, copy=False)
        deme_sizes = np.unique(md["deme"], return_counts=True)
        if as_dict is False:
            return deme_sizes
        return {i: j for i, j in zip(deme_sizes[0], deme_sizes[1])}

    def dump_tables_to_tskit(
        self,
        *,
        model_params: Optional[Union[ModelParams,
                                     Dict[str, ModelParams]]] = None,
        demes_graph: Optional[demes.Graph] = None,
        population_metadata: Optional[Dict[int, object]] = None,
        data: Optional[object] = None,
        seed: Optional[int] = None,
        parameters: Optional[Dict] = None,
        destructive=False,
    ) -> tskit.TreeSequence:
        """
        Dump the population's TableCollection into
        an tskit TreeSequence

        :param model_params: Model parameters to be stored as top-level metadata
        :type model_params: :class:`fwdpy11.ModelParams` or :class:`dict`

        :param demes_graph: A demographic model specified via `demes`.
        :type demes_graph: :class:`demes.Graph`

        :param population_metadata: A mapping from integer id of a deme/population to metadata
        :type population_metadata: dict

        :param data: User-generated data to add to top-level metadata
        :type data: object

        :param seed: Random number seed for top-level metadata.
                     If not `None`, `seed` must be `>= 0`.
        :type seed: int

        :param parameters: The simulation parameters for the provenance table.
        :type parameters: None or dict

        :param destructive: If `True`, delete data held by the current instance.
        :type destructive: bool

        :rtype: tskit.TreeSequence

        For examples, see :ref:`tskitconvert_vignette`.

        .. warning::

            If `destructive` is `True`, further opertations on the
            current instance should be considered undefined behavior
            that may lead to a crash.

        .. versionchanged:: 0.8.2

            Added `parameters`.
            Generate provenance information for return value.
            The provenance information is validated using
            :func:`tskit.validate_provenance`, which may
            raise an exception.

        .. versionchanged:: 0.10.0

            Use tskit metadata schema.
            Mutation time is now stored in the tskit.MutationTable column.
            Origin time of mutations is part of the metadata.

        .. versionchanged:: 0.14.0

            Added `destructive` option.

        .. versionchanged:: 0.15.0

            Added `model_params`, `demes_graph`, `population_metadata`,
            `data` keyword args.

        .. versionchanged:: 0.15.2

            Fixed bug that could generate a :class:`tskit.PopulationTable`
            with an incorrect number of rows.

        .. versionchanged:: 0.17.0

            The `wrapped` keyword argument is deprecated.

        .. versionadded:: 0.19.0

            Removed deprecated `wrapped` keyword argument.

        """
        return fwdpy11.tskit_tools._dump_tables_to_tskit._dump_tables_to_tskit(
            self,
            model_params=model_params,
            demes_graph=demes_graph,
            population_metadata=population_metadata,
            data=data,
            seed=seed,
            parameters=parameters,
            destructive=destructive,
        )

    def dump_to_file(self, filename: str):
        """
        Write a population to a file in binary format.
        """
        self._dump_to_file(filename)

    def pickle_to_file(self, filename: IO):
        """
        Pickle the population to an open file.

        This function may be preferred over
        the direct pickling method because it uses less
        memory.  It is, however, slower.

        To read the population back in, you must call
        :func:`fwdpy11.DiploidPopulation.load_from_pickle_file`.

        :param f: A handle to an open file

        .. versionadded:: 0.3.0

        """
        self._pickle_to_file(filename)

    def sample_timepoints(
        self, include_alive=True
    ):
        """
        Return an iterator over all sample time points.
        The iterator yields time, nodes, and metadata.

        :param include_alive: If True, include currently-alive individuals.

        .. versionadded :: 0.5.1

        .. versionchanged:: 0.5.3

           Monkey-patched into pybind11 class
        """
        amd = np.array(self.ancient_sample_metadata, copy=False)
        amd.flags.writeable = False
        times, utimes = self._get_times()
        for uti in utimes:
            mdidx = np.where(times == uti)[0]
            sample_nodes = amd["nodes"][mdidx].flatten()
            sample_nodes.flags.writeable = False
            mdslice = amd[mdidx][:]
            mdslice.flags.writeable = False
            yield uti, sample_nodes, mdslice

        if include_alive is True:
            md = np.array(self.diploid_metadata, copy=False)
            nodes = md["nodes"].flatten()
            md.flags.writeable = False
            nodes.flags.writeable = False
            yield self.generation, nodes, md

    def add_mutation(
        self,
        rng: fwdpy11.GSLrng,
        *,
        window: Tuple[float, float] = None,
        ndescendants: int = 1,
        deme: Optional[int] = None,
        data: NewMutationData = None,
    ) -> Optional[int]:
        """
        Add a new mutation to the population.

        .. versionadded:: 0.16.0

        :param rng: Random number generator
        :type rng: fwdpy11.GSLrng

        The following arguments are keyword-only:

        :param window: A window [left, right) within which to place the mutation.
                       The default is `None`, meaning the window is the entire
                       genome.
        :type window: tuple[float, float]
        :param ndescendants: The number of alive nodes carrying the new mutation.
                             Default is `1`, implying that a singleton mutation
                             will be generated.
        :type ndescendants: int
        :param deme: The deme in which to place the new mutation
                     The default is `None`, meaning that alive node demes are
                     not considered.
        :type deme: int
        :type data: The details of the new mutation
        :type data: fwdpy11.NewMutationData

        :returns: The key of the new mutation.
                  (The index of the variant in :attr:`DiploidPopulation.mutations`.)

        Implementation details:

        * A set of nodes are found within `window` that are ancestral to
          exactly `ndescendants` alive nodes.
        * From this list of candidate nodes, one is chosen randomly
          to be the node where we place the new mutation.
        * If the node's parent's time is NULL, then the mutations' origin
          time is the first 64 bit integer ancestral to the node time,
          which could be the node time itself.
          Otherwise, a 64 bit integer is chosen uniformly between the
          node time and the node's parent's time.
        * If `deme` is None, any set of `ndescendants` will be considered.
        * If `deme` is :math:`\geq 0`, all `ndescendants` alive nodes must be
          from that deme.
        * If the `deme`/`ndescendants` requirements cannot be satisified,
          the function returns `None`.
        * The genetic values of individuals are not updated by this function.
          Such updates will happen when the population begins evolution
          forwards in time.

        Exceptions:

        :raises ValueError: for bad input.
        :raises RuntimeError: if this function is applied during a simulation
        :raises RuntimeError: if internal checks fail.
        """
        if window is None:
            _window = (0.0, self.tables.genome_length)
        else:
            _window = window

        if deme is None:
            _deme = -1
        else:
            _deme = deme

        if ndescendants is None:
            raise ValueError(f"ndescendants must be > 0: got {ndescendants}")

        from fwdpy11._fwdpy11 import _add_mutation

        key = _add_mutation(
            rng, _window[0], _window[1], ndescendants, _deme, data, self
        )
        if key == np.iinfo(np.uint64).max:
            return None
        return key
